### DISCUSSION TITLE
NoClassDefFoundError org/mapstruct/factory/Mappers with OpenJDK 11 - SpringBoot App

### DISCUSSION THREAD

**Discussion by rishabhvarma09** [mapstruct/mapstruct#2714] [Q&A]
I am working on the SpringBoot application which has two instances, one is using Java 8, and another one is with Java 11. Both applications are exactly similar in terms of design structure.

**Case #1: Application with Java 8 + Spring + Lombok**

In the above application with Java 8, I am only using **mapstruct-processor dependency 1.2.0.Final** with maven-compiler-plugin having a configuration as **-Amapstruct.defaultComponentModel=spring** in pom file and we are able to retrieve different mappers using both approaches **The Mapper Factory - Mappers.getMapper(XYZ.class)** and **Using Dependency Injection - @Autowired** at the same time.

**Note: For above, mapstruct-1.2.0.Final.jar and mapstruct-processor-1.2.0.Final.jar both JARs are getting downloaded** 

**Case #2: Application with Java 11 + Spring + Lombok**

In the above application with Java 11, after adding **mapstruct-processor dependency 1.2.0.Final** with maven-compiler-plugin having a configuration as **-Amapstruct.defaultComponentModel=spring** in pom file. It is not downloading **mapstruct-1.2.0.Final.jar** this time and I need to explicitly add **mapstruct dependency 1.2.0.Final**. Also, I need to forcefully add either **mapstruct-processor dependency 1.2.0.Final** or **annotationProcessorPaths** in the maven-compiler-plugin otherwise my application is not able to compile.

**Problem with Java 11 application:** When I am adding both mapstruct and mapstruct-processor in pom it is no longer able to retrieve different mappers using both approaches **The Mapper Factory - Mappers.getMapper(XYZ.class)** and **Using Dependency Injection - @Autowired** at the same time. It is throwing NoClassDefFoundError org/mapstruct/factory/Mappers exception at the time of Mappers.getMapper() execution.

StackTrace:
exception [Handler dispatch failed; nested exception is java.lang.NoClassDefFoundError: org/mapstruct/factory/Mappers] with root cause java.lang.ClassNotFoundException: org.mapstruct.factory.Mappers
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)

PS: I know that it has been suggested to not use Mapper Factory with Spring but at the moment it is not really possible for me to change the structural design of my application while migrating to Java 11. I am also interested to understand why different behaviors occurred with Java 8 and 11 although both the applications are able to generate Impl classes in the generated source. Really appreciate any suggestions for a workaround, please 🤞

**Comment by Zegveld** [comment]
I do not have a quick answer to this, but seeing the symptoms I believe that the mapstruct module is missing from the dependencies of the application in the java 11 version.
Can you check if your maven setup conforms to what is listed here: https://mapstruct.org/documentation/installation/
